codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customizations/SigV4EventStreamSupportStructures.kt

```diff
@@ -0,0 +1,313 @@
+                                        }
+                                    },
+                                    Err(err) => Err(err),
+                                }
+                            }
+                            Err(extraction_err) => Ok(#{UnmarshalledMessage}::Error(#{SignedEventError}::InvalidSignedEvent(extraction_err))),
+                        }
+                    }
+                }
+                """,
+                "UnmarshallMessage" to
+                    CargoDependency.smithyEventStream(runtimeConfig).toType()
+                        .resolve("frame::UnmarshallMessage"),
+                "UnmarshalledMessage" to
+                    CargoDependency.smithyEventStream(runtimeConfig).toType()
+                        .resolve("frame::UnmarshalledMessage"),
+                "Message" to CargoDependency.smithyTypes(runtimeConfig).toType().resolve("event_stream::Message"),
+                "EventStreamError" to CargoDependency.smithyEventStream(runtimeConfig).toType().resolve("error::Error"),
+                "SignedEvent" to signedEvent(runtimeConfig),
+                "SignedEventError" to signedEventError(runtimeConfig),
+                "extract_signed_message" to extractSignedMessage(runtimeConfig),
+                *RuntimeType.preludeScope,
+            )
+        }
+
+    private fun extractSignedMessage(runtimeConfig: RuntimeConfig): RuntimeType =
+        RuntimeType.forInlineFun("extract_signed_message", supportModule) {
+            rustTemplate(
+                """
+                /// Result of extracting a potentially signed message
+                ##[derive(Debug)]
+                pub enum MaybeSignedMessage {
+                    /// Message was signed and has been extracted
+                    Signed {
+                        /// The inner message that was signed
+                        message: #{Message},
+                        /// Signature information from the outer message
+                        signature: #{SignatureInfo},
+                    },
+                    /// Message was not signed (no `:chunk-signature` header present)
+                    Unsigned,
+                }
+
+                /// Extracts the inner message from a potentially signed event stream message.
+                pub fn extract_signed_message(message: &#{Message}) -> #{Result}<MaybeSignedMessage, #{ExtractionError}> {
+                    // Check if message has chunk signature
+                    let mut chunk_signature = None;
+                    let mut timestamp = None;
+
+                    for header in message.headers() {
+                        match header.name().as_str() {
+                            ":chunk-signature" => {
+                                if let #{HeaderValue}::ByteArray(bytes) = header.value() {
+                                    chunk_signature = Some(bytes.as_ref().to_vec());
```

Comment:
Are there limits on event stream message/chunk sizes to prevent memory exhaustion attacks? or is this the service's responsibility via Hyper configuration? Are individual header values size-limited? or should we enforce some check here that the value is maximum 256 bytes long before we `.to_vec()` it?

Reply:
the message has been long-parsed before this point. `Message` isn't a streaming object.